controllers/sample_controller_rendered_resources.go

```diff
@@ -0,0 +1,359 @@
+				rateLimiter.Frequency,
+				rateLimiter.Burst,
+			),
+		}).
+		Complete(r)
+}
+
+// TemplateRateLimiter implements a rate limiter for a client-go.workqueue.  It has
+// both an overall (token bucket) and per-item (exponential) rate limiting.
+func TemplateRateLimiter(failureBaseDelay time.Duration, failureMaxDelay time.Duration,
+	frequency int, burst int,
+) ratelimiter.RateLimiter {
+	return workqueue.NewMaxOfRateLimiter(
+		workqueue.NewItemExponentialFailureRateLimiter(failureBaseDelay, failureMaxDelay),
+		&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(frequency), burst)})
+}
+
+// Reconcile is the entry point from the controller-runtime framework.
+// It performs a reconciliation based on the passed ctrl.Request object.
+func (r *SampleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	sampleObj := v1alpha1.Sample{}
+
+	if err := r.Client.Get(ctx, req.NamespacedName, &sampleObj); err != nil {
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		logger.Info(req.NamespacedName.String() + " got deleted!")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// verify resource interface compliance
+	objectInstance := *sampleObj.DeepCopy()
+
+	// check if deletionTimestamp is set, retry until it gets deleted
+	status := getStatusFromObjectInstance(&objectInstance)
+
+	// set state to Deleting if not set for an object with deletion timestamp
+	if !objectInstance.GetDeletionTimestamp().IsZero() &&
+		status.State != v1alpha1.StateDeleting {
+		// if the status is not yet set to deleting, also update the status
+		return ctrl.Result{}, r.setStatusForObjectInstance(ctx, &objectInstance, status.WithState(v1alpha1.StateDeleting))
+	}
+
+	// add finalizer if not present
+	if controllerutil.AddFinalizer(&objectInstance, finalizer) {
+		objectInstance.SetManagedFields(nil)
+		return ctrl.Result{}, r.ssa(ctx, &objectInstance)
+	}
+
+	switch status.State {
+	case "":
+		return ctrl.Result{}, r.HandleInitialState(ctx, &objectInstance)
+	case v1alpha1.StateProcessing:
+		return ctrl.Result{Requeue: true}, r.HandleProcessingState(ctx, &objectInstance)
+	case v1alpha1.StateDeleting:
+		return ctrl.Result{Requeue: true}, r.HandleDeletingState(ctx, &objectInstance)
+	case v1alpha1.StateError:
+		return ctrl.Result{Requeue: true}, r.HandleProcessingState(ctx, &objectInstance)
+	case v1alpha1.StateReady:
+		return ctrl.Result{RequeueAfter: requeueInterval}, r.HandleReadyState(ctx, &objectInstance)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// HandleInitialState bootstraps state handling for the reconciled resource.
+func (r *SampleReconciler) HandleInitialState(ctx context.Context, objectInstance *v1alpha1.Sample) error {
+	status := getStatusFromObjectInstance(objectInstance)
+
+	installationCondition := metav1.Condition{
```

Comment:
this condition will not be set to false if we are already in error state. I suggest to move this logic outside of initial state handling.